flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/stream/sql/ProcessTableFunctionTest.java

```diff
@@ -0,0 +1,326 @@
+        util.addTemporarySystemFunction("f", TableAsSetFunction.class);
+        assertReachesOptimizer("SELECT * FROM f(r => TABLE t1 PARTITION BY name, i => 1)");
+    }
+
+    @Test
+    void testTableAsSetWithInvalidPartitionBy() {
+        util.addTemporarySystemFunction("f", TableAsSetFunction.class);
+        assertThatThrownBy(() -> util.verifyRelPlan("SELECT * FROM f(r => TABLE t1, i => 1)"))
+                .hasRootCauseMessage(
+                        "Table argument 'r' requires a PARTITION BY clause for parallel processing.");
+    }
+
+    @Test
+    void testTableAsSetOptionalPartitionBy() {
+        util.addTemporarySystemFunction("f", TableAsSetOptionalPartitionFunction.class);
+        assertReachesOptimizer("SELECT * FROM f(r => TABLE t1, i => 1)");
+    }
+
+    @Test
+    void testTypedTableAsSet() {
+        util.addTemporarySystemFunction("f", TypedTableAsSetFunction.class);
+        assertReachesOptimizer("SELECT * FROM f(u => TABLE t1 PARTITION BY name, i => 1)");
+    }
+
+    @Test
+    void testTypedTableAsSetWithInvalidInput() {
+        util.addTemporarySystemFunction("f", TypedTableAsSetFunction.class);
+        // function expects <STRING name, INT score>
+        assertThatThrownBy(
+                        () ->
+                                util.verifyRelPlan(
+                                        "SELECT * FROM f(u => TABLE t3 PARTITION BY name, i => 1)"))
+                .hasMessageContaining(
+                        "No match found for function signature "
+                                + "f(<RecordType(CHAR(3) name, BOOLEAN isValid)>, <NUMERIC>, <CHARACTER>)");
+    }
+
+    @Test
+    void testEmptyArgs() {
+        util.addTemporarySystemFunction("f", EmptyArgFunction.class);
+        util.verifyRelPlan("SELECT * FROM f(uid => 'my-ptf')");
+    }
+
+    @Test
+    void testPojoArgs() {
+        util.addTemporarySystemFunction("f", PojoArgsFunction.class);
+        util.addTemporarySystemFunction("pojoCreator", PojoCreatingFunction.class);
+        assertReachesOptimizer(
+                "SELECT * FROM f(input => TABLE t1, scalar => pojoCreator('Bob', 12), uid => 'my-ptf')");
+    }
+
+    @Test
+    void testInvalidTableFunction() {
+        util.addTemporarySystemFunction("f", NoProcessTableFunction.class);
+        assertThatThrownBy(() -> util.verifyRelPlan("SELECT * FROM f(r => TABLE t1)"))
+                .hasRootCauseMessage(
+                        "Only scalar arguments are supported at this location. "
+                                + "But argument 'r' declared the following traits: [TABLE, TABLE_AS_ROW]");
+    }
+
+    @Test
+    void testReservedArg() {
+        util.addTemporarySystemFunction("f", ReservedArgFunction.class);
+        assertThatThrownBy(() -> util.verifyRelPlan("SELECT * FROM f(uid => 'my-ptf')"))
+                .hasRootCauseMessage(
+                        "Function signature must not declare system arguments. Reserved argument names are: [uid]");
+    }
```

Comment:
Not sure how easy it is for `success` cases
however at least for `fail` cases wouldn't it make sense to replace them with a couple of parameterized tests?